source/Halibut/Transport/SecureListener.cs

```diff
@@ -183,6 +194,68 @@ void WaitForPendingConnectionOrCancellation()
             }
         }
 
+        async Task AcceptAsyncIgnoringExceptions(CancellationToken cancellationToken)
+        {
+            try
+            {
+                await AcceptAsync(cancellationToken);
+            }
+            catch (Exception)
+            {
+                // This matches what we did before, in theory this will never happen.
+            }
+        }
+        async Task AcceptAsync(CancellationToken cancellationToken)
+        {
+
+            const int errorThreshold = 3;
+            
+            // Don't call listener.Stop until we sure we are not doing an Accept
+            // See: https://github.com/OctopusDeploy/Issues/issues/6035
+            // See: https://github.com/dotnet/corefx/issues/26034
+            using (IsWindows() ? cancellationToken.Register(listener.Stop) : (IDisposable)null!)
+            {
+                var numberOfFailedAttemptsInRow = 0;
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    TcpClient? client = null;
+                    try
+                    {
+#if !NETFRAMEWORK
+            client = await listener.AcceptTcpClientAsync(this.cancellationToken);
```

Comment:
I don't think I understand the subtlety here.

In the old one, we would wait until `cts.IsCancellationRequested || listener.Pending()` before calling accept.

Using the cancellation token removes the need for `cts.IsCancellationRequested`. But is waiting for `listener.Pending()` important? Or does this happen for free somehow with using the async versions of `AcceptTcpClientAsync`?

Reply:
The whole wait until we know something is pending was done to avoid sitting in Accept at the same time as listener.stop() being called, because a bug did (or still does honestly I can't tell) where the call to listener.stop() would hang if AcceptTcpClient was in the middle of accepting a new connection.


The big change is to remove the SpinWait that `polls listener.pending()` because that is very expensive :(.
Instead we use the new (relative to when the old code was added) method that is `AcceptTcpClientAsync` with a cancellation token. When we shut down on linux the order is always:
1. Cancel the CTS
2. Wait for the loop to exit (which it does)
3. then and only then stop the listener (at this point we know we are NOT in accept and so can not hang).